### ISSUE TITLE
optimizing complex objective function using T.grad

### ISSUE THREAD

**Issue by Imorton-zd** [Theano/Theano#3537]
My research tast needs optimizing an complex objective function:
![y5cyd 0e980bwy qw d s](https://cloud.githubusercontent.com/assets/13865086/10625741/5d5feb10-77d9-11e5-840a-9dc784e8cd19.png)
q, a* and a are matrixes all given. Now I want to update wq, bq, wa and ba. How can I using T.grad to get the partial derivative of wq, bq, wa and ba. Opinions/Views would be highly appreciated!

**Comment by Imorton-zd** [comment]
@mgermain @abergeron @dwf @nouiz @lamblin Could you spend your valuable time solving this question? Thanks a lot.

**Comment by lamblin** [comment]
The functionality of gradient wrt complex variables (the objective still has to be a scalar real, so minimization makes sense) is unfortunately not supported in Theano, and would require some amount of work (there is a proposal at http://deeplearning.net/software/theano/proposals/complex_gradient.html).
Unless you would want to tackle that, the easier option for now would be to express every operation on real variables, and consider real and imaginary parts of your numbers as different variables, or an additional dimension (for instance, an (m, n) complex matrix could be represented as an (m, n, 2) tensor).

**Comment by lucainnocenti** [comment]
@mgermain @lamblin are there any fundamental technical difficulties in properly implementing the gradient of a complex cost with respect to a real variable? Would this be relatively trivial but time consuming or would it require a fundamental restructuring of the code? (see also a [related question on stackoverflow](http://stackoverflow.com/q/41193260/4063051))

**Comment by lamblin** [comment]
Backpropagating the gradient of a real scalar wrt complex tensors is something that could be done. Most of the effort would be to implement the gradient for all the Ops needing it, while making sure unsupported Ops do not silently return something wrong.

But I'm not sure that the gradient of a complex cost makes sense, or is properly defined mathematically. Moreover, how do you minimize a complex cost? Even if we figure out a way, if there are intermediate complex variables, that means the backpropagation needs to handle 4 real derivatives (real / imaginary wrt. real / imaginary parts), which cannot be stored in complex, and may be hard to express efficiently as one operation.

**Comment by lucainnocenti** [comment]
@lamblin indeed, I understand the difficulty in properly defining and using the gradient wrt complex tensors. My question is however on the implementation of the gradient of complex functions, *wrt to a real variable*. A trivial example of this would be the gradient of x1 + 2j * x` wrt `x`, giving the complex quantity `2j`. While this case work, more complex cases easily fail, as shown for example in the linked StackOverflow post

**Comment by lamblin** [comment]
The gradient of complex functions wrt real variables is not easier to define.
In particular in the backpropagation when intermediate variables are complex (even if, ultimately, the end variable is real).
Don't forget that in your case, your gradient wrt a real is complex. What would be the gradient wrt an intermediate complex variable? Many of the functions we use are not analytic.

**Comment by lucainnocenti** [comment]
@lamblin ok I see what you mean about the necessity to derive "wrt to complex variables". I'm not sure I see how that is much of a (conceptual) problem, or why you think the analyticity is that relevant.

In general you can always think of a complex function `f(z)` as a real function with two (real) inputs and two (real) outputs: `f(z) = (f_R(z_R, z_I), f_I(z_R, z_I))`, so that the derivative can always be just thought to be wrt to real variables: real and imaginary parts of the inputs. Can you give an example in which you think this kind of gradient wouldbe problematic to define?

**Comment by lamblin** [comment]
In the general case, we would need to store and propagate 4 (real) tensors for each operation: df_R/dz_R, df_R/dz_I, df_I/dz_R and df_I/dz_I.

This makes things conceptually harder because we do not have a data type to store those 4 numbers easily. It makes it practically harder in general to have one operation that computes these 4 expressions at the same time in an efficient way (and not one after the other, reading the same array multiple times). Memory efficiency may also be an issue if those 4 values could be represented in a more compact way (if some are all zeros, or some are easily expressed from the others).

If the function were analytic, we would have df_R/dz_R = df_I/dz_I and df_R/dz_I = - df_I/dz_R, so we would only need to store 2 (real) tensors, which could be done in a complex tensor, and those can usually be computed by complex primitives (complex add / mul, zdot, ...).

Having a real starting point (cost for backprop, or it could be the input evaluation point for the forward mode) also enables to only care about 2 (real) tensors, which we can store in a complex tensor, and hopefully compute at once via complex primitives.
